packages/tpetra/core/test/ImportExport2/AsyncTransfer_UnitTests.cpp

```diff
@@ -693,6 +780,246 @@ namespace {
   }
 
 
+  template <typename Packet, typename LO, typename GO>
+  class CheckDefaultTransferStatus {
+  private:
+    using DistObjectRCP = RCP<DistObject<Packet, LO, GO>>;
+
+  public:
+    CheckDefaultTransferStatus(FancyOStream& o, bool& s)
+      : out(o),
+        success(s)
+    { }
+
+    void operator()(DistObjectRCP source, DistObjectRCP target) const {
+      TEST_ASSERT(target->transferArrived());
+    }
+
+  private:
+    FancyOStream& out;
+    bool& success;
+  };
+
+  TEUCHOS_UNIT_TEST_TEMPLATE_3_DECL( TransferArrived, MultiVector_defaultTrue, Scalar, LO, GO )
+  {
+    MultiVectorTransferFixture<Scalar, LO, GO> fixture(out, success);
+
+    fixture.setup(0);
+    fixture.performTransfer(CheckDefaultTransferStatus<Scalar, LO, GO>(out, success));
+  }
+
+  TEUCHOS_UNIT_TEST_TEMPLATE_3_DECL( TransferArrived, CrsMatrix_defaultTrue, Scalar, LO, GO )
+  {
+    DiagonalCrsMatrixTransferFixture<Scalar, LO, GO> fixture(out, success);
+
+    fixture.setup();
+    fixture.performTransfer(CheckDefaultTransferStatus<char, LO, GO>(out, success));
+  }
+
+
+  template <typename Packet, typename LO, typename GO>
+  class CheckTransferArrivedForwardImport {
+  private:
+    using DistObjectRCP = RCP<DistObject<Packet, LO, GO>>;
+
+  public:
+    CheckTransferArrivedForwardImport(FancyOStream& o, bool& s)
+      : out(o),
+        success(s)
+    { }
+
+    void operator()(DistObjectRCP source, DistObjectRCP target) const {
+      Import<LO, GO> importer(source->getMap(), target->getMap());
+
+      target->beginImport(*source, importer, INSERT);
+      while (!target->transferArrived());
```

Comment:
Is this while loop calling transferArrived needed? 
Is the intent of this test to determine whether transferArrived is true after endImport/endExport?

Reply:
it's not strictly needed for the workflow -- the endImport will wait until everything has arrived anyway, and the main point of the test is that transferArrived is still true after endImport/Export.

The loop does represent a subtle assert that transferArrived does in fact return true at some point after beginImport and before endImport.  i.e., we want to TEST_ASSERT(target->transferArrived()) after the begin import (kind of the opposite check as the one in the "transfer not arrived" test), but we can't call the assert directly because we don't know exactly when things arrived.  So we loop as a way of checking that it does at some point become true.

Since there's not an explicit assert here, the fact that this loop actually represents an assert is probably too subtle, but I wasn't sure how else to do it.  I'm also open to an argument for just dropping it if it's too confusing.  Thoughts?